### PR TITLE
fix(rerouting): check that the new route is different

### PR DESCRIPTION
**Pull request by lilnasy**
## Changes
- Fixes closed issue #8257
- The issue has to do with implicit rerouting which will be deprecated and replaced with an [explicit API](https://github.com/withastro/roadmap/discussions/681) but that may take a while.

## Testing
Added 4 cases based on what I saw people encounter in the issue and on discord.

## Docs
Does not affect usage.

### CODE REVIEW

**Review by ematipico — APPROVED** [review]
Code looks good to me. I wonder if we can add a test case, the same we have in the issue

**Comment by lilnasy** [comment]
Let's put a potential endless loop in our CI 🥳

File: packages/astro/src/vite-plugin-astro-server/route.ts [hunk 1]
```diff
@@ -220,6 +220,7 @@ export async function handleRoute({
 	let response = await pipeline.renderRoute(renderContext, mod);
 	if (response.status === 404 && has404Route(manifestData)) {
 		const fourOhFourRoute = await matchRoute('/404', manifestData, pipeline);
+		// if (fourOhFourRoute?.route !== options.route)
```
Comment (lilnasy) [inline packages/astro/src/vite-plugin-astro-server/route.ts:223]: Commenting out the fix to see what happens to CI.

Comment (lilnasy) [reply packages/astro/src/vite-plugin-astro-server/route.ts]: CI times out. Logs for `astro:test` [never appear](https://github.com/withastro/astro/actions/runs/6292593299/job/17092598249?pr=8648#step:9:747). It wouldn't be possible to tell if the cases failed.

**Comment by lilnasy** [comment]
They seem to be synchronous, so there's not stopping them.

**Comment by ematipico** [comment]
Well, let's not overthink it then. You tested manually :)

**Comment by lilnasy** [comment]
And we don't have the bug in our codebase anymore! 🥳

_at least this particular instance of it_

**Review by matthewp — APPROVED** [review]

**Pull request merged**